aten/src/ATen/cuda/CachingHostAllocator.cpp

```diff
@@ -72,6 +74,173 @@ struct CUDACachingHostAllocatorImpl
+    if (!CachingHostAllocatorImpl<CUDAStream, EventPool::Event>::record_event(ptr, ctx, s)) {
+      return false;
+    }
+
+    // Inspect the preceding captured node on stream s. If it's a
+    // memcpy from pinned host memory that reads from this block, hash
+    // the copied section and record it to detect accidental mutations
+    // during capture. The purpose is to identify cases where the CPU
+    // writes to host pinned memory after the GPU is scheduled to read
+    // from it during cuda graph replay. I am calling this a
+    // CPU-Write-After-GPU-Read hazard. Such a situation is almost
+    // certainly a bug, bcause it doesn't match the semantics of eager
+    // launch. That check happens in end_allocate_to_pool.
+
+    Block *block = nullptr;
+    if (block_exists(ctx)) {
+      block = reinterpret_cast<Block*>(ctx);
+    } else {
+      block = get_block_from_ptr(ptr);
+    }
+
+    assert(block != nullptr);
+
+    CUDAStream stream(CUDAStream::UNCHECKED, s);
+
+    std::optional<std::tuple<const cudaGraphNode_t*, size_t>> terminals_opt = c10::cuda::streamGetTerminalNodes(stream);
+
+    if (!terminals_opt.has_value()) {
+      // no stream capture is happening right now
+      return true;
+    }
+
+    auto&& [terminals, num_terminals] = *terminals_opt;
+
+    if (num_terminals == 0) {
+      TORCH_WARN(
+          "CachingHostAllocator: no nodes precede record_event() during stream capture; "
+          "this is probably a bug.");
+      return true;
+    }
+
+    if (num_terminals > 1) {
+      TORCH_WARN(
+          "CachingHostAllocator: multiple preceding nodes during capture; "
+          "skipping host read verification.");
+      return true;
+    }
+
+    cudaGraphNode_t node = terminals[0];
+    cudaGraphNodeType type{cudaGraphNodeTypeCount};
+    C10_CUDA_CHECK(cudaGraphNodeGetType(node, &type));
+    if (type != cudaGraphNodeTypeMemcpy) {
+      TORCH_WARN(
+          "CachingHostAllocator: preceding captured node is not a memcpy; "
+          "can only check for CPU-write-after-GPU-read errors for memcpy nodes.");
+      return true;
+    }
+
+    cudaMemcpy3DParms params{};
+    C10_CUDA_CHECK(cudaGraphMemcpyNodeGetParams(node, &params));
+
+    // Only handle Host->Device copies (source is host memory)
+    if (params.kind != cudaMemcpyHostToDevice && params.kind != cudaMemcpyDefault) {
+      // don't warn, since Device->Host copy is a valid node to precede record_event()
+      return true;
+    }
+
+    cudaPointerAttributes attr{};
+    cudaError_t attr_err = cudaPointerGetAttributes(&attr, params.srcPtr.ptr);
+    // TODO: Think about what to do if the memory type is
+    // cudaMemoryTypeUnregistered, which I believe refers to pageable
+    // host memory. (Meanwhile, cudaMemoryTypeHost represents pinned
+    // host memory.)
+    if (attr_err != cudaSuccess || attr.type != cudaMemoryTypeHost) {
+      // Clear error if any and return (not pinned host memory)
+      (void)cudaGetLastError();
```

Comment:
Use [maybe unused] in C++17, the void trick is unnecessary now.

Although I think there is a different c10 function if all you want to do is clear an error. Shouldn't the error at least be logged anyway?